### PR TITLE
Add various new fields to Primo

### PR DESCRIPTION
**Pull request by adam3smith**
- improve call numbers
- add full text URLs
- add links to finding aids
- some minor parsing improvements for publishers and contributors
- update tests

@zuphilip -- the folks at Harvard University Library requested these changes (the item type and field ones, not the parsing) and told me they made sense across Primo instances. If you could have a quick look that all this looks right to you? (note that there is some Harvard specific code)

@cbaksik -- FYI

### CODE REVIEW

File: Primo.js [hunk 1]
```diff
@@ -369,9 +384,51 @@ function importPNX(text) {
...
+	}
+	if (callArray.length) {
+		item.callNumber = callArray.join(", ");
+	}
+	else {
+		ZU.xpathText(doc, '//enrichment/classificationlcc');
+	}
 	
+	//Harvard specific code, requested by Harvard Library:
+	var library = ZU.xpathText(doc, '//browse/institution')
```
Comment (zuphilip) [inline Primo.js]: Missing semicolon (also in lines 83 and 303)

File: Primo.js [hunk 2]
```diff
@@ -265,8 +278,10 @@ function importPNX(text) {
 	}
 	
 	item.title = ZU.xpathText(doc, '//display/title');
-	if(item.title) item.title = ZU.unescapeHTML(item.title);
-	
+	if(item.title) {
+		item.title = ZU.unescapeHTML(item.title);
+	    item.title = item.title.replace(/\s*:/, ":");
```
Comment (zuphilip) [inline Primo.js]: Is this for separating a subtitle? And fix indenting here.

File: Primo.js [hunk 3]
```diff
@@ -369,9 +384,51 @@ function importPNX(text) {
...
+	for (var i = 0; i<callNumber.length; i++) {
+		if (callNumber[i].textContent.search(/\$\$D.+\$/) != -1) {
+			callArray.push(callNumber[i].textContent.match(/\$\$D(.+?)\$/)[1]);
+		}
+	}
+	if (!callArray.length) {
+		callNumber = ZU.xpath(doc, '//display/availlibrary');
+		for (var i = 0; i<callNumber.length; i++) {
+			if (callNumber[i].textContent.search(/\$\$2.+\$/) != -1) {
+				callArray.push(callNumber[i].textContent.match(/\$\$2\(?(.+?)(?:\s*\))?\$/)[1]);
```
Comment (zuphilip) [inline Primo.js]: Here is an example for us:
```
<availlibrary>
   $$IMAN$$L$$1geschlossenes Magazin$$2(ZL 474 )$$Savailable$$X49MAN_INST$$YWEST$$ZT-MAG$$P1
</availlibrary>
<availlibrary>
   $$IMAN$$L$$1geschlossenes Magazin$$2(ZL 474 )$$Savailable$$X49MAN_INST$$YWEST$$ZT-MAG$$P2
</availlibrary>
```
So, yes, I guess that should work.

**Comment by zuphilip** [comment]
I cannot test and look much further at the moment, but the changes in general look good for me to. Actually, they will also be an improvement for us (call number and links). Thus, thank you very much for the work!

File: Primo.js (hunk 2 above)
Comment (cbaksik) [inline Primo.js]: It's for removing a single space in front of a semicolon. In library cataloging, titles and subtitles are separate by a space-semicolon-space, but in citation styles, the space in front of the semicolon needs to be removed, and this change will prevent users from having to do it manually in Zotero

File: Primo.js (hunk 3 above)
Comment (adam3smith) [inline Primo.js]: this looks to me like I should be removing duplicates in the array before converting it to a string. Will do that.

File: Primo.js [hunk 4]
```diff
@@ -504,11 +562,12 @@ var testCases = [
...
 				],
 				"date": "1973",
-				"ISBN": "0600393046",
+				"ISBN": "9780600393047",
+				"callNumber": "9B6655",
 				"language": "eng",
 				"libraryCatalog": "Primo",
 				"numPages": "252",
-				"publisher": "Hamlyn",
+				"publisher": "Hamlyn,",
```
Comment (adam3smith) [inline Primo.js]: note to self: check where that comma is coming from and remove.

File: Primo.js (hunk 2 above)
Comment (adam3smith) [inline Primo.js]: we already do this for import via MARC and from WorldCat, so this should be entirely uncontroversial.

**Comment by adam3smith** [comment]
@cbaksik - this is now live with Zotero (translators update every 24hs or use the "Update Now" button in the general tab of the preferences) & ready for testing. Feedback here or via e-mail.

File: Primo.js (hunk 2 above)
Comment (zuphilip) [inline Primo.js]: Yeah, I was just wondering... BTW I think the space-semicolon-space comes from the [ISBD](https://de.wikipedia.org/wiki/International_Standard_Bibliographic_Description) format.

**Pull request merged**